kube-runtime/src/controller/mod.rs

```diff
@@ -348,16 +349,24 @@ where
     /// with a configurable `Context`.
     pub fn run<ReconcilerFut, T>(
         self,
-        reconciler: impl FnMut(K, Context<T>) -> ReconcilerFut,
+        mut reconciler: impl FnMut(K, Context<T>) -> ReconcilerFut,
         error_policy: impl FnMut(&ReconcilerFut::Error, Context<T>) -> ReconcilerAction,
         context: Context<T>,
     ) -> impl Stream<Item = Result<(ObjectRef<K>, ReconcilerAction), Error<ReconcilerFut::Error, watcher::Error>>>
     where
         K: Clone + Meta + 'static,
-        ReconcilerFut: TryFuture<Ok = ReconcilerAction>,
-        ReconcilerFut::Error: std::error::Error + 'static,
+        ReconcilerFut: TryFuture<Ok = ReconcilerAction> + Send + 'static,
+        ReconcilerFut::Error: std::error::Error + Send + 'static,
     {
-        applier(reconciler, error_policy, context, self.reader, self.selector)
+        applier(
+            move |obj, ctx| {
+                CancelableJoinHandle::spawn(reconciler(obj, ctx).into_future(), &Handle::current())
+            },
```

Comment:
The idea is that we want to cancel the reconciler if we cancel the applier, rather than letting it leak (and continue running forever in the background).